src/rt/sections_elf_shared.d

```diff
@@ -466,7 +468,17 @@ extern(C) void _d_dso_registry(CompilerDSOData* data)
 
         freeDSO(pdso);
 
-        if (_loadedDSOs.empty) finiLocks(); // last DSO
+        if (_loadedDSOs.empty)
+        {
+            // last DSO
+            version (Shared)
+            {
+                !pthread_mutex_lock(&_handleToDSOMutex) || assert(0);
+                _handleToDSO.reset();
+                !pthread_mutex_unlock(&_handleToDSOMutex) || assert(0);
```

Comment:
Druntime cannot get unloaded while being used, so locking isn't necessary here.
An assertion that the hash table is empty would be in order as well.